design/workflow.go

```diff
@@ -0,0 +1,207 @@
+		Example("An MLFlow project codeset")
+	})
+	Field(3, "type", String, "The type of the input (codeset, string, ...)", func() {
+		Example("codeset")
+	})
+	Field(4, "default", String, "Default value for the input", func() {
+		Example("mlflow-example")
+	})
+	Field(5, "labels", ArrayOf(String), "Labels associated with the input", func() {
+		Example([]string{"mlflow-project"})
+	})
+})
+
+// WorkflowOutput defines the output from a FuseML workflow
+var WorkflowOutput = Type("WorkflowOutput", func() {
+	Field(1, "name", String, "Name of the output", func() {
+		Example("prediction-url")
+	})
+	Field(2, "description", String, "Description of the output", func() {
+		Example("The URL where the exposed prediction service endpoint can be contacted to run predictions.")
+	})
+	Field(3, "type", String, "The data type of the output", func() {
+		Example("string")
+	})
+})
+
+// WorkflowStep defines a step for a FuseML workflow
+var WorkflowStep = Type("WorkflowStep", func() {
+	Field(1, "name", String, "The name of the step", func() {
+		Example("predictor")
+	})
+	Field(2, "image", String, "The image used to execute the step", func() {
+		Example("ghcr.io/fuseml/kfserving-predictor:1.0")
+	})
+	Field(3, "inputs", ArrayOf(WorkflowStepInput), "List of inputs for the step")
+	Field(4, "outputs", ArrayOf(WorkflowStepOutput), "List of output from the step")
+	Field(5, "env", ArrayOf(StepEnv), "List of environment variables available for the container running the step")
+})
+
+// WorkflowStepInput defines the input for a FuseML workflow step
+var WorkflowStepInput = Type("WorkflowStepInput", func() {
+	Field(1, "name", String, "Name of the input", func() {
+		Example("model-uri")
+	})
+	Field(2, "value", String, "Value of the input", func() {
+		Example("s3://mlflow-artifacts/3/c7ae3b0e6fd44b4b96f7066c66672551/artifacts/model")
+	})
+	Field(3, "codeset", StepInputCodeset, "Codeset associated with the input")
+})
+
+// StepInputCodeset defines the Codeset type of input for a FuseML workflow step
+var StepInputCodeset = Type("StepInputCodeset", func() {
+	Field(1, "name", String, "Name or ID of the codeset", func() {
+		Example("mlflow-project")
+	})
+	Field(2, "path", String, "Path where the codeset will be mounted inside the container running the step", func() {
+		Example("/project")
+	})
+})
+
+// WorkflowStepOutput defines the output from a FuseML workflow step
+var WorkflowStepOutput = Type("WorkflowStepOutput", func() {
+	Field(1, "name", String, "Name of the variable to hold the step output value", func() {
+		Example("model-uri")
+	})
+	Field(2, "image", StepOutputImage, "If the step builds a container image as output it will be referenced as 'image'")
+})
+
+// StepOutputImage defines the output from a FuseML workflow when it builds a container image
+var StepOutputImage = Type("StepOutputImage", func() {
+	Field(1, "dockerfile", String, "Path to the Dockerfile used to build the image", func() {
```

Comment:
I asked about this before and did not get it then. I still do not get it now: why would the "Dockerfile used to build the image" be an argument of the output structure for building an image?

Reply:
I believe the idea is that as the step outputs an image, the `dockerfile` defines how to build that image (output).

Reply:
Exactly, this is yet another type of input/output artifact supported by fuseml for pipeline steps (and runnables too, which reminds me that I'll need to include this in the runnables API). The "image" artifact may need more fields in addition to `dockerfile` and `name`. For example, the image contents also need to come from somewhere, so maybe a list of container paths will also be required, but let's worry about that later.